modules/drivers/video/tools/decode_video/frame_processor.cc

```diff
@@ -0,0 +1,116 @@
+
+#include <fstream>
+#include <memory>
+#include <sstream>
+
+#include "cyber/common/log.h"
+#include "modules/drivers/video/tools/decode_video/h265_decoder.h"
+
+namespace apollo {
+namespace drivers {
+namespace video {
+
+FrameProcessor::FrameProcessor(const std::string& input_video_file,
+                               const std::string& output_jpg_dir)
+    : output_jpg_dir_(output_jpg_dir) {
+  std::ifstream video_file(input_video_file, std::ios::binary);
+  std::vector<uint8_t> video_buffer(
+      (std::istreambuf_iterator<char>(video_file)),
+      std::istreambuf_iterator<char>());
+  input_video_buffer_ = video_buffer;
+}
+
+bool FrameProcessor::ProcessStream() const {
+  if (input_video_buffer_.empty()) {
+    AERROR << "error: failed to read from input video file";
+    return false;
+  }
+  AVCodecParserContext* codec_parser = av_parser_init(AV_CODEC_ID_H265);
+  if (codec_parser == nullptr) {
+    AERROR << "error: failed to init AVCodecParserContext";
+    return false;
+  }
+  AVPacket apt;
+  av_init_packet(&apt);
+  const std::unique_ptr<H265Decoder> decoder(new H265Decoder());
+  if (!decoder->Init()) {
+    AERROR << "error: failed to init decoder";
+    return false;
+  }
+  uint32_t local_size = static_cast<uint32_t>(input_video_buffer_.size());
+  uint8_t* local_data = const_cast<uint8_t*>(input_video_buffer_.data());
+  AINFO << "decoding: video size = " << local_size;
+  int frame_num = 0;
+  int error_frame_num = 0;
+  while (local_size > 0) {
+    int frame_len = av_parser_parse2(
+        codec_parser, decoder->GetCodecCtxH265(), &(apt.data), &(apt.size),
+        local_data, local_size, AV_NOPTS_VALUE, AV_NOPTS_VALUE, AV_NOPTS_VALUE);
+    if (apt.data == nullptr) {
+      apt.data = local_data;
+      apt.size = local_size;
+    }
+    AINFO << "frame " << frame_num << ": frame_len=" << frame_len
+          << ". left_size=" << local_size;
+    std::vector<uint8_t> jpeg_buffer = decoder->Process(apt.data, apt.size);
+    if (!jpeg_buffer.empty()) {
+      WriteOutputJpgFile(jpeg_buffer, GetOutputFile(frame_num));
+      frame_num++;
+    } else {
+      error_frame_num++;
+    }
+    local_data += frame_len;
+    local_size -= frame_len;
+  }
+  // Trying to decode the left over frames from buffer
+  for (int i = error_frame_num; i >= 0; i--) {
+    std::vector<uint8_t> jpeg_buffer = decoder->Process(nullptr, 0);
+    if (!jpeg_buffer.empty()) {
+      WriteOutputJpgFile(jpeg_buffer, GetOutputFile(frame_num));
+      AINFO << "frame " << frame_num << ": read from buffer";
+      frame_num++;
+    }
+  }
+  AINFO << "total frames: " << frame_num;
+  av_parser_close(codec_parser);
+  return true;
+}
+
+const std::string FrameProcessor::GetOutputFile(const int frame_num) const {
+  const int SUFFIX_LEN = 5;
+  std::stringstream jpg_suffix;
+  jpg_suffix.fill('0');
+  jpg_suffix.width(SUFFIX_LEN);
+  jpg_suffix << std::to_string(frame_num);
+  return output_jpg_dir_ + "/" + jpg_suffix.str() + ".jpg";
```

Comment:
Try "apollo::common::util::StrCat", which concat pieces together and avoid making many intermediate string objects.